### DISCUSSION TITLE
spacy fails on import (CBlas DLL load fails)

### DISCUSSION THREAD

**Discussion by cb-air** [explosion/spaCy#11591] [Help: Installation] [answered]
I'm trying to get spacy loaded, but it fails as shown below.  The problem persists after reinstalling both thinc and spacy in the global site-packages directory.  
```
>>> import thinc
>>> import spacy
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Program Files\Python37\lib\site-packages\spacy\__init__.py", line 6, in <module>
    from .errors import setup_default_warnings
  File "C:\Program Files\Python37\lib\site-packages\spacy\errors.py", line 2, in <module>
    from .compat import Literal
  File "C:\Program Files\Python37\lib\site-packages\spacy\compat.py", line 38, in <module>
    from thinc.api import Optimizer  # noqa: F401
  File "C:\Program Files\Python37\lib\site-packages\thinc\api.py", line 2, in <module>
    from .initializers import normal_init, uniform_init, glorot_uniform_init, zero_init
  File "C:\Program Files\Python37\lib\site-packages\thinc\initializers.py", line 4, in <module>
    from .backends import Ops
  File "C:\Program Files\Python37\lib\site-packages\thinc\backends\__init__.py", line 7, in <module>
    from .ops import Ops
  File "C:\Program Files\Python37\lib\site-packages\thinc\backends\ops.py", line 16, in <module>
    from .cblas import CBlas
ImportError: DLL load failed: The specified module could not be found.
```
[EDIT: to replace screenshot with text]

**Comment by polm** [comment, marked as answer]
Sorry you're having trouble installing spaCy, but please don't share images of code or error output, they're hard to read and use. Please share text by copy and pasting it. 

It looks like you have failed to install CBlas somehow, which is weird because it should just be included with Thinc. You generally shouldn't have to install Thinc manually if you're using spaCy. How did you install spaCy? I would recommend going through the [install guide](https://spacy.io/usage) if you haven't. 

Some things to check:

- what versions of spaCy and Thinc do you have? (use `pip freeze` to check)
- are you using 32 bit Python? (run this in Python: `import sys;print("%x" % sys.maxsize, sys.maxsize > 2**32)`)

**Comment by cb-air** [comment]
Thanks for your help. 

spacy==3.4.1
spacy-legacy==3.0.10
spacy-loggers==1.0.3
thinc==8.1.2

`7fffffffffffffff True`

They were installed without error using:
```
py -m pip install thinc
py -m pip install spacy
```

**Reply by polm** [reply]
Sorry for the delayed reply. That output all looks OK. 

You should have a file with a name similar to this:

```
C:\Program Files\Python37\lib\site-packages\thinc\backends\cblas.cp37-win_amd64.pyd
```

Can you check that such a file exists? The name may be a bit different, but let me know if you find anything similar. The main thing is that it starts with `cblas` and ends with `pyd`. It could be a problem if that file doesn't exist, or if it's not the 64 bit version. Those things shouldn't happen, but sometimes they do. 

It may also help to install spaCy in a virtualenv, rather than directly in your global Python environment. (Also, I assume `py` is just a shortcut, but if it's a wrapper or something please let us know.)

**Reply by cb-air** [reply]
Yes, the cblas .pyd file you reference exists and is the 64-bit version. 

The `py` is a shortcut. I'm actually in a virtual environment now. You can see the exact install syntax and output below. 

I have 2 ideas I am pursuing, but would welcome thoughts:
  
  1. Look for dependencies for the .pyd file and put those DLL into the same directory
       see for instance:
https://es.stackoverflow.com/questions/547357/c%C3%B3mo-soluciono-este-problema-con-los-dll-en-python-3-7-9-shell
https://stackoverflow.com/questions/8846480/error-importing-a-pyd-file-as-a-python-module-from-a-pyo-file (second bullet of top answer says to run depends.exe on the missing cblas DLL, and copy those dependencies to the same folder.)

  2. Installing/Updating Visual C++ install (it's currently 2013 version)

Thank you for the assistance. 

```
(.venv) C:\\cb\pii_removal>pip install -U pip setuptools wheel
Requirement already satisfied: pip in C:\\cb\pii_removal\.venv\lib\site-packages (22.2.2)
Requirement already satisfied: setuptools in C:\\cb\pii_removal\.venv\lib\site-packages (47.1.0)
Collecting setuptools
  Using cached setuptools-65.4.1-py3-none-any.whl (1.2 MB)
Collecting wheel
  Using cached wheel-0.37.1-py2.py3-none-any.whl (35 kB)
Installing collected packages: wheel, setuptools
  Attempting uninstall: setuptools
    Found existing installation: setuptools 47.1.0
    Uninstalling setuptools-47.1.0:
      Successfully uninstalled setuptools-47.1.0
Successfully installed setuptools-65.4.1 wheel-0.37.1

(.venv) C:\\cb\pii_removal>pip install -U spacy
Collecting spacy
  Using cached spacy-3.4.1-cp37-cp37m-win_amd64.whl (12.0 MB)

[snip]

Installing collected packages: wasabi, cymem, zipp, urllib3, typing-extensions, spacy-loggers, spacy-legacy, smart-open, murmurhash, MarkupSafe, langcodes,
idna, colorama, charset-normalizer, certifi, blis, tqdm, requests, pydantic, preshed, jinja2, importlib-metadata, catalogue, srsly, click, typer, confection, thinc, pathy, spacy
  Attempting uninstall: typing-extensions
    Found existing installation: typing_extensions 4.4.0
    Uninstalling typing_extensions-4.4.0:
      Successfully uninstalled typing_extensions-4.4.0
Successfully installed MarkupSafe-2.1.1 blis-0.7.8 catalogue-2.0.8 certifi-2022.9.24 charset-normalizer-2.1.1 click-8.1.3 colorama-0.4.5 confection-0.0.3 cymem-2.0.6 idna-3.4 importlib-metadata-5.0.0 jinja2-3.1.2 langcodes-3.3.0 murmurhash-1.0.8 pathy-0.6.2 preshed-3.0.7 pydantic-1.9.2 requests-2.28.1 smart-open-5.2.1 spacy-3.4.1 spacy-legacy-3.0.10 spacy-loggers-1.0.3 srsly-2.4.4 thinc-8.1.4 tqdm-4.64.1 typer-0.4.2 typing-extensions-4.1.1 urllib3-1.26.12 wasabi-0.10.1 zipp-3.9.0

(.venv) C:\\cb\pii_removal>python -m spacy download en_core_web_sm
Traceback (most recent call last):
  File "C:\Program Files\Python37\lib\runpy.py", line 183, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "C:\Program Files\Python37\lib\runpy.py", line 142, in _get_module_details
    return _get_module_details(pkg_main_name, error)
  File "C:\Program Files\Python37\lib\runpy.py", line 109, in _get_module_details
    __import__(pkg_name)
  File "C:\\cb\pii_removal\.venv\lib\site-packages\spacy\__init__.py", line 6, in <module>
    from .errors import setup_default_warnings
  File "C:\\cb\pii_removal\.venv\lib\site-packages\spacy\errors.py", line 2, in <module>
    from .compat import Literal
  File "C:\\cb\pii_removal\.venv\lib\site-packages\spacy\compat.py", line 38, in <module>
    from thinc.api import Optimizer  # noqa: F401
  File "C:\\cb\pii_removal\.venv\lib\site-packages\thinc\api.py", line 2, in <module>
    from .initializers import normal_init, uniform_init, glorot_uniform_init,
zero_init
  File "C:\\cb\pii_removal\.venv\lib\site-packages\thinc\initializers.py", line 4, in <module>
    from .backends import Ops
  File "C:\\cb\pii_removal\.venv\lib\site-packages\thinc\backends\__init__.py", line 7, in <module>
    from .ops import Ops
  File "C:\\cb\pii_removal\.venv\lib\site-packages\thinc\backends\ops.py", line 16, in <module>
    from .cblas import CBlas
ImportError: DLL load failed: The specified module could not be found.

[snip]

Directory of C:\\cb\pii_removal\.venv\Lib\site-packages\thinc\backends

10/12/2022  04:30 PM    <DIR>          .
10/12/2022  04:30 PM    <DIR>          ..
10/12/2022  04:30 PM                 0 __init__.pxd
10/12/2022  04:30 PM             5,584 __init__.py
10/12/2022  04:30 PM             2,246 _cupy_allocators.py
10/12/2022  04:30 PM            18,902 _custom_kernels.cu
10/12/2022  04:30 PM            22,708 _custom_kernels.py
10/12/2022  04:30 PM             5,398 _murmur3.cu
10/12/2022  04:30 PM             2,821 _param_server.py
10/12/2022  04:30 PM           110,080 cblas.cp37-win_amd64.pyd
10/12/2022  04:30 PM           774,279 cblas.cpp
10/12/2022  04:30 PM             1,396 cblas.pxd
10/12/2022  04:30 PM             1,104 cblas.pyx
10/12/2022  04:30 PM            15,083 cpu_kernels.hh
10/12/2022  04:30 PM            12,059 cupy_ops.py
10/12/2022  04:30 PM            32,256 linalg.cp37-win_amd64.pyd
10/12/2022  04:30 PM           247,884 linalg.cpp
10/12/2022  04:30 PM             8,643 linalg.pxd
10/12/2022  04:30 PM                57 linalg.pyx
10/12/2022  04:30 PM               607 mps_ops.py
10/12/2022  04:30 PM           512,000 numpy_ops.cp37-win_amd64.pyd
10/12/2022  04:30 PM         2,604,992 numpy_ops.cpp
10/12/2022  04:30 PM             2,039 numpy_ops.pxd
10/12/2022  04:30 PM            40,181 numpy_ops.pyx
10/12/2022  04:30 PM            59,848 ops.py
10/12/2022  04:30 PM    <DIR>          __pycache__
```

**Reply by polm** [reply]
Hm, that all looks fine, so I'm not really sure what could be going on... One unfortunate thing about the Windows API is that when it gives a DLL error like this, the issue could be with the directly loaded DLL, or any one of its dependencies. So the advice about copying the DLLs is worth a shot. I think the CBLAS contained in Thinc should be self contained, and not have any `.dll` file, but maybe there is some OS dependcy it loads. 

I would definitely try upgrading Visual C++. Usually if that's an issue there's a very clear error telling you to upgrade, but maybe something weird is going on related to it?

Anyway, let me know how those ideas go, and I'll think about it and let you know if I come up with anything!

**Reply by alanbr94** [reply]
Hello! I had the same issue and tried all the steps. Worked updating visual c++. Thanks!

**Reply by polm** [reply]
Thanks for the update, that's great to hear! 

@cb-air Let us know if upgrading works for you.

**Reply by cb-air** [reply]
Yes, updating visual c++ was the solution. Thanks @alanbr94 for the input, and thank you very much @polm for the ongoing, friendly help. Much appreciated!

**Reply by polm** [reply]
Thanks for confirming! It's unfortunate the errors weren't more clear in this situation, but now we know what to do next time it comes up.

**Comment by tushar2242** [comment]
raceback (most recent call last):
  File "C:\Users\test\Desktop\keyhash\keyword_extractor\app.py", line 2, in <module>
    from keyword_tag import Take_text
  File "C:\Users\test\Desktop\keyhash\keyword_extractor\keyword_tag.py", line 1, in <module>
    import spacy
  File "C:\Users\test\Desktop\keyhash\keyword_extractor\Lib\site-packages\spacy\__init__.py", line 6, in <module>
    from .errors import setup_default_warnings
  File "C:\Users\test\Desktop\keyhash\keyword_extractor\Lib\site-packages\spacy\errors.py", line 3, in <module>
    from .compat import Literal
  File "C:\Users\test\Desktop\keyhash\keyword_extractor\Lib\site-packages\spacy\compat.py", line 39, in <module>
    from thinc.api import Optimizer  # noqa: F401
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\test\Desktop\keyhash\keyword_extractor\Lib\site-packages\thinc\api.py", line 1, in <module>
    from .backends import (
  File "C:\Users\test\Desktop\keyhash\keyword_extractor\Lib\site-packages\thinc\backends\__init__.py", line 17, in <module>
    from .cupy_ops import CupyOps
  File "C:\Users\test\Desktop\keyhash\keyword_extractor\Lib\site-packages\thinc\backends\cupy_ops.py", line 16, in <module>
    from .numpy_ops import NumpyOps
ImportError: DLL load failed while importing numpy_ops: The specified module could not be found.


annotated-types==0.6.0
blinker==1.7.0
blis==0.7.11
catalogue==2.0.10
certifi==2024.2.2
charset-normalizer==3.3.2
click==8.1.7
cloudpathlib==0.16.0
colorama==0.4.6
confection==0.1.4
cymem==2.0.8
Flask==3.0.2
idna==3.6
itsdangerous==2.1.2
Jinja2==3.1.3
joblib==1.3.2
langcodes==3.3.0
MarkupSafe==2.1.5
murmurhash==1.0.10
numpy==1.26.4
packaging==23.2
preshed==3.0.9
pydantic==2.6.1
pydantic_core==2.16.2
requests==2.31.0
scikit-learn==1.4.1.post1
scipy==1.12.0
setuptools==69.1.0
smart-open==6.4.0
spacy==3.7.4
spacy-legacy==3.0.12
spacy-loggers==1.0.5
srsly==2.4.8
thinc==8.2.3
threadpoolctl==3.3.0
tqdm==4.66.2
typer==0.9.0
typing_extensions==4.9.0
urllib3==2.2.1
wasabi==1.1.2
weasel==0.3.4
Werkzeug==3.0.1
wheel==0.42.0
